packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts

```diff
@@ -0,0 +1,450 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import * as dirPath from "path";
+import { mkdirSync } from "fs";
+import { strict as assert } from "assert";
+import {
+    BaseFuzzTestState,
+    createFuzzDescribe,
+    createWeightedGenerator,
+    Generator,
+    interleave,
+    makeRandom,
+    performFuzzActions,
+    Reducer,
+    SaveInfo,
+    take,
+} from "@fluid-internal/stochastic-test-utils";
+import {
+    MockContainerRuntimeFactoryForReconnection,
+    MockContainerRuntimeForReconnection,
+    MockFluidDataStoreRuntime,
+    MockStorage,
+} from "@fluidframework/test-runtime-utils";
+import { IChannelServices } from "@fluidframework/datastore-definitions";
+import { DirectoryFactory, SharedDirectory } from "../../directory";
+import { IDirectory } from "../../interfaces";
+
+interface Client {
+    sharedDirectory: SharedDirectory;
+    containerRuntime: MockContainerRuntimeForReconnection;
+}
+
+interface FuzzTestState extends BaseFuzzTestState {
+    containerRuntimeFactory: MockContainerRuntimeFactoryForReconnection;
+    clients: Client[];
+}
+
+interface ClientSpec {
+    directoryId: string;
+}
+
+interface SetKey extends ClientSpec {
+    type: "set";
+    path: string;
+    key: string;
+    value: string;
+}
+
+interface ClearKeys extends ClientSpec {
+    type: "clear";
+    path: string;
+}
+
+interface DeleteKey extends ClientSpec {
+    type: "delete";
+    path: string;
+    key: string;
+}
+
+interface CreateSubDirectory extends ClientSpec {
+    type: "createSubDirectory";
+    path: string;
+    name: string;
+}
+
+interface DeleteSubDirectory extends ClientSpec {
+    type: "deleteSubDirectory";
+    path: string;
+    name: string;
+}
+
+interface Synchronize {
+    type: "synchronize";
+}
+
+type KeyOperation = SetKey | DeleteKey | ClearKeys;
+
+type SubDirectoryOperation = CreateSubDirectory | DeleteSubDirectory;
+
+type ClientOperation = KeyOperation | SubDirectoryOperation;
+
+type Operation = ClientOperation | Synchronize;
+
+interface OperationGenerationConfig {
+    validateInterval: number,
+    maxSubDirectoryChild?: number;
+    subDirectoryNamePool?: string[];
+    keyNamePool?: string[];
+}
+
+const defaultOptions: Required<OperationGenerationConfig> = {
+    validateInterval: 2,
+    maxSubDirectoryChild: 3,
+    subDirectoryNamePool: ["dir1", "dir2", "dir3"],
+    keyNamePool: ["prop1", "prop2", "prop3"],
+};
+
+function makeOperationGenerator(optionsParam?: OperationGenerationConfig): Generator<Operation, FuzzTestState> {
+    const options = { ...defaultOptions, ...(optionsParam ?? {})};
+    type ClientOpState = FuzzTestState & { sharedDirectory: SharedDirectory; };
+
+    // All subsequent helper functions are generators; note that they don't actually apply any operations.
+    function pickAbsolutePathForCreateDirectoryOp(state: ClientOpState): string {
```

Comment:
it feels like you could probably simplify the `pick` methods here by writing a generic method to pick a subdirectory and rejection sample it based on the op (e.g. for subdir create, reject if the selected directory is already at max size).

Reply:
I combined the methods for generating key ops. There seems to be a good diff between the dir ops one, so keeping them separate for now.